### PR TITLE
2.38 compatibility test fixes

### PR DESCRIPTION
**Pull request by deeonwuli**
### :pushpin: References

-   **Issue:** Closes https://app.clickup.com/t/3k6naef

### CODE REVIEW

**Review by nshandra — APPROVED** [review]
Looks good.

In my test between DHIS v2.36 and 2.38 the metadata POST response format changed from:
```
{
    "importParams": {
        "userOverrideMode": "NONE",
        "importMode": "COMMIT",
        "identifier": "UID",
        "preheatMode": "REFERENCE",
        "importStrategy": "CREATE_AND_UPDATE",
        "atomicMode": "ALL",
        "mergeMode": "REPLACE",
        "flushMode": "AUTO",
        "skipSharing": false,
        "skipTranslation": false,
        "skipValidation": false,
        "metadataSyncImport": false,
        "firstRowIsHeader": true,
        "username": "admin"
    },
    "status": "OK",
    "stats": {
        "created": 0,
        "updated": 0,
        "deleted": 0,
        "ignored": 0,
        "total": 0
    },
    "typeReports": []
}
```
to:
```
{
    "httpStatus": "OK",
    "httpStatusCode": 200,
    "status": "OK",
    "response": {
        "responseType": "ImportReport",
        "status": "OK",
        "stats": {
            "created": 0,
            "updated": 0,
            "deleted": 0,
            "ignored": 0,
            "total": 0
        },
        "typeReports": []
    }
}
```
So, the new interface allows using either version.

**Comment by anagperal** [comment]
I'm not able to update any sharing settings (tested dashboard, program and dataset in https://dev.eyeseetea.com/who-dev-238, https://dev.eyeseetea.com/play and in https://play.dhis2.org/2.38.6 with merge and replace). In any api response publicAccess, userAccesses or userGroupAccesses it's obtained. Because of that, in ApplySharingSettingsUseCase any payload is sent to update, we always send in the POST:
```
{
  "dataSets": [],
  "categoryCombos": [],
  "categoryOptionCombos": [],
  "categories": [],
  "categoryOptions": [],
  "dataElements": []
}
```
This could be because of this: https://docs.dhis2.org/es_419/develop/using-the-api/dhis-core-version-238/sharing.html#new-sharing-object. Now we have the sharing object so we need to add this logic too


image

File: src/domain/entities/MetadataItem.ts [hunk 1]
```diff
@@ -55,12 +55,18 @@ export interface DataDimensionItem {
     programIndicator?: Ref;
 }
 
-export type MetadataItem = Ref & SharedObject & { [key: string]: any | undefined };
+export type MetadataItem = Ref & {
+    sharing: {
```
Comment (anagperal) [inline src/domain/entities/MetadataItem.ts]: Create here a new type:

```typescript
export type Sharing = {
    sharing: SharingObject;
};

export type SharingObject = {
    userGroups: Record<string, SharingSetting>;
    users: Record<string, SharingSetting>;
    public: string;
};

export type MetadataItem = Ref & Sharing & SharedObject & { [key: string]: any | undefined };
```

File: src/domain/usecases/metadata/ApplySharingSettingsUseCase.ts [hunk 2]
```diff
@@ -37,15 +37,17 @@ export class ApplySharingSettingsUseCase implements UseCase {
     }
 
     private updateSharingSettings(item: MetadataItem, sharings: SharedObject, replace: boolean): MetadataItem {
```
Comment (anagperal) [inline src/domain/usecases/metadata/ApplySharingSettingsUseCase.ts:39]: You need also to update `userGroups`, `users` and `public` inside new `sharing` object in every item (not only `publicAccess`, `userAccesses` and `userGroupAccesses` that are ok). In addition, you need to apply the function `assertValidSharingSettings` to `acces`  property inside properties in the new `sharing` object

**Review by anagperal — CHANGES_REQUESTED** [review]
thanks @deeonwuli ! I've added some comments (functional testing and code checked)

**Comment by anagperal** [comment]
In addition to this comments, in SelectMetadataStep and in ListDependenciesStep components when rows are created add this to each item for displaying in the table the info in the columns publicAccess, userAccesses, userGroupAccesses
```typescript
publicAccess: item.publicAccess ?? item.sharing.public,
userGroupAccesses: item.userGroupAccesses ?? Object.values(item.sharing.userGroups),
userAccesses: item.userAccesses ?? Object.values(item.sharing.users),
```

File: src/domain/entities/MetadataItem.ts [hunk 3]
```diff
@@ -55,12 +55,18 @@ export interface DataDimensionItem {
...
+    };
+} & SharedObject & { [key: string]: any | undefined };
 
 export function isValidModel(model: string): model is MetadataModel {
     return ["dataSets", "programs", "dashboards"].includes(model);
 }
 
 export function isValidMetadataItem(item: any): item is MetadataItem {
-    return item.id && item.publicAccess && item.userAccesses && item.userGroupAccesses;
+    return item.id && item.sharing;
```
Comment (anagperal) [inline src/domain/entities/MetadataItem.ts]: To make it retro-compatible:  item should have or sharing object or the other properties

File: src/domain/entities/SharedObject.ts [hunk 4]
```diff
@@ -1,8 +1,6 @@
-import { NamedRef } from "./Ref";
+import { SharingRule } from "@eyeseetea/d2-ui-components";
```
Comment (anagperal) [inline src/domain/entities/SharedObject.ts]: You never have to take entities of your domain from outside, instead create a new one:
```typescript
export type SharingSetting = { id: Id; displayName: string; access: string; name: string };

```

**Comment by anagperal** [comment]
thanks @deeonwuli for all the fixes! It works :tada:! Only one comment about `SharingSetting` entity and that's all!
Screenshot from 2024-04-26 08-02-55

**Comment by nshandra** [comment]
Given the review from Ana i will remove myself from the reviewers.

**Review by anagperal — APPROVED** [review]
@deeonwuli thanks!!

**Pull request merged**